src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_keychain.c

```diff
@@ -224,6 +224,105 @@ static OSStatus DeleteInKeychain(CFTypeRef needle, SecKeychainRef haystack)
     return status;
 }
 
+static bool IsCertInKeychain(CFTypeRef needle, SecKeychainRef haystack)
+{
+    CFMutableDictionaryRef query = CFDictionaryCreateMutable(
+        kCFAllocatorDefault, 0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
+
+    if (query == NULL)
+        return errSecAllocate;
+
+    const void* constHaystack = haystack;
+    CFTypeRef result = NULL;
+    CFArrayRef searchList = CFArrayCreate(NULL, &constHaystack, 1, &kCFTypeArrayCallBacks);
+
+    if (searchList == NULL)
+    {
+        CFRelease(query);
+        return errSecAllocate;
+    }
+
+    CFArrayRef itemMatch = CFArrayCreate(NULL, (const void**)(&needle), 1, &kCFTypeArrayCallBacks);
+
+    if (itemMatch == NULL)
+    {
+        CFRelease(searchList);
+        CFRelease(query);
+        return errSecAllocate;
+    }
+
+    CFDictionarySetValue(query, kSecReturnRef, kCFBooleanTrue);
+    CFDictionarySetValue(query, kSecMatchSearchList, searchList);
+    CFDictionarySetValue(query, kSecMatchItemList, itemMatch);
+    CFDictionarySetValue(query, kSecClass, kSecClassCertificate);
+    OSStatus status = SecItemCopyMatching(query, &result);
+
+    bool ret = true;
+
+    if (status == errSecItemNotFound)
+    {
+        ret = false;
+    }
+
+    if (status == noErr)
+    {
+        CFDictionarySetValue(query, kSecClass, kSecClassIdentity);
+        status = SecItemCopyMatching(query, &result);
+    }
```

Comment:
This looks weird. Should not this be in `status == errSecItemNotFound`? After all the purpose is to try to find either a certificate or an identity...

Reply:
I agree. I've modified the check so that in case it's not found as a cert we always check if it's an identity.